# Borrower reliability analysis

The customer is the credit department of the bank. It is necessary to find out whether the marital status and the number of children of the client affect the fact of repaying the loan on time. Input data from the bank - statistics on the solvency of customers.
The results of the study will be taken into account when building a credit scoring model - a special system that assesses the ability of a potential borrower to return a loan to a bank.

**The purpose of the study** is to test four hypotheses:
1. Repayment of the loan on time depends on the number of children.
2. Repayment of the loan on time depends on marital status.
3. Repayment of the loan on time depends on the level of income.
4. Different purposes of the loan affect its repayment on time.

**Research Progress**

The study will take place in three stages:
 1. Data review.
 2. Data preprocessing.
 3. Hypothesis testing.

## Data overview

Display the first fifteen rows of the table:

In [37]:
import pandas as pd # pandas library import /datasets/data.csv

In [38]:
df = pd.read_csv('C:/Users/tatar/downloads/data.csv') # reading data file and saving to df
df.head(15) # getting the first 15 rows of a df table

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   

     total_income                            purpose  
0   253875.639453                      покупка жилья  
1   112080.014102            приобретение автомобиля  
2   145885.952297                      покупка жилья  
3   267628.550329         дополнительное образование  
4   158616.077870                    сыграть свадьбу  
5   255763.565419                      покупка жилья  
6   240525.971920                  операции с жильем  
7   135823.934197                        образование  
8    95856.832424              на проведение свадьбы  
9   144425.938277            покупка жилья для семьи  
10  113943.491460               покупка недвижимости  
11   77069.234271  покупка коммерческой недвижимости  
12            NaN                    сыграть свадьбу  
13  130458.228857            приобретение автомобиля  
14  165127.911772         покупка жилой недвижимости

Get general information about the table.

In [39]:
df.info() # получение общей информации о данных в таблице df

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


The table has twelve columns. The data types in the columns are different - float64, int64, object.

According to the data documentation:

* `children` — the number of children in the family;
* `days_employed` — total work experience in days;
* `dob_years` — client's age in years;
* `education` — the level of education of the client;
* `education_id` — education level identifier;
* `family_status` — marital status;
* `family_status_id` — marital status identifier;
* `gender` — gender of the client;
* `income_type` — employment type;
* `debt` — whether he had a debt to repay loans;
* `total_income` — monthly income;
* `purpose` — the purpose of the loan.

**Conclusions**

Each row of the table contains data on borrowers. Some of the columns describe financial data: the presence of debt, monthly income and the purpose of obtaining a loan. The rest of the data tells about the client: the presence of children, work experience, age, education, marital status, gender and type of employment.

Preliminarily, it can be argued that the data are sufficient to test the hypotheses. But there are incorrect data (some values of seniority are negative, information about education is written in violation of style) and there are gaps in the data.

Problems in the data need to be fixed.

## Data preprocessing

### Fill in missing values

First, let's count the number of missing values in the table.

In [40]:
df.isna().sum() # missing values count

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Missing values were found in the `days_employed` (total days of employment) and `total_income` (monthly income) columns. The reason for the appearance of gaps in the data is most likely that the clients did not want to provide this data for some reason. 

In [41]:
print('The proportion of missing values in the data is:', round(2174 / 21525 * 100, 2), '%')

The proportion of missing values in the data is: 10.1 %


Since the remaining values in the columns with gaps are presented in a wide range and there is always the possibility of the presence of too large values that affect the distribution of data in the sample, we will replace the gaps with the median value.

Let's take into account that both the total length of service in days and the monthly income directly depend on the type of employment. Therefore, we will replace the gaps in the table with median values for the `type of employment` group.

In [42]:
for income in df['income_type'].unique(): # forming a cycle by unique values of the type of employment
    median_1 = df.loc[df['income_type'] == income, 'days_employed'].median() # calculating median for column 'days_employed'
    median_2 = df.loc[df['income_type'] == income, 'total_income'].median() # calculating median for column 'total_income'
    df.loc[(df['days_employed'].isna()) & (df['income_type'] == income), 'days_employed'] = median_1 # change of missing values
    df.loc[(df['total_income'].isna()) & (df['income_type'] == income), 'total_income'] = median_2 # change of missing values

Make sure there are no missing values in the table. To do this, once again calculate the missing values.

In [43]:
df.isna().sum() # missing values count

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

And once again display the first fifteen rows of the table:

In [44]:
df.head(15) # getting the first 15 rows of a df table

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0  365213.306266         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   

     total_income                            purpose  
0   253875.639453                      покупка жилья  
1   112080.014102            приобретение автомобиля  
2   145885.952297                      покупка жилья  
3   267628.550329         дополнительное образование  
4   158616.077870                    сыграть свадьбу  
5   255763.565419                      покупка жилья  
6   240525.971920                  операции с жильем  
7   135823.934197                        образование  
8    95856.832424              на проведение свадьбы  
9   144425.938277            покупка жилья для семьи  
10  113943.491460               покупка недвижимости  
11   77069.234271  покупка коммерческой недвижимости  
12  118514.486412                    сыграть свадьбу  
13  130458.228857            приобретение автомобиля  
14  165127.911772         покупка жилой недвижимости

### Checking data for anomalies and corrections

For better readability of the table, we will round the values in the `total_income` column to two decimal places.

In [45]:
df['total_income'] = round(df['total_income'], 2) # rounding values to two decimal places
df.head() # check

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0     253875.64   
1   женат / замужем                 0      F   сотрудник     0     112080.01   
2   женат / замужем                 0      M   сотрудник     0     145885.95   
3   женат / замужем                 0      M   сотрудник     0     267628.55   
4  гражданский брак                 1      F   пенсионер     0     158616.08   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

Let's check the `children` column for data anomalies.

In [46]:
df['children'].value_counts() # checking for data anomalies in the 'children' column

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Two problems are identified here. Negative value, and too large a value. Both errors are most likely caused by banal "mistakes". For a negative value (-1), we will overwrite that column's data using a modulo function. For a value that is too large (20), we will replace 20 with 2, since both the mean and median of the list of numbers from 0 to 5 are the same and equal to 2.5.

In [47]:
df['children'] = abs(df['children']) # replacing negative values in 'children' column with module
df['children'] = df['children'].replace(20, 2) # replacing value '20' with '2'
df['children'].value_counts() # check

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

Let's check the `dob_years` column for anomalies.

In [48]:
df['dob_years'].value_counts() # data anomaly check in 'dob_years' column

35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
22    183
66    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

It can be seen that 101 clients have 0 in the age column, this can be either a reluctance of clients to indicate their age or a technical error. Let's replace the zeros with the median value.

In [49]:
df['dob_years'] = df['dob_years'].replace(0, df['dob_years'].median()) # replacing zero values with medians
df['dob_years'].value_counts() # check

42    698
35    617
40    609
41    607
34    603
38    598
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
22    183
66    183
67    167
21    111
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

Let's process the anomalies in the data on seniority, for some clients they are displayed in a negative range. This may be due to incorrect filling of the table either manually, or problems when importing from other sources.

In [50]:
df['days_employed'] = abs(df['days_employed']) # replacing negative values in 'days_employed' column with module

Abnormally high data on seniority were also noticed, it is necessary to process them. To begin with, we will determine from what value the length of service will be abnormally large.

According to statistics in the Russian Federation over the past 5 years, there were an average of 247 working days a year, taking into account the annual leave of 28 days (per year) required by labor legislation, let's take 220 working days a year. It is also necessary to take into account the limitation of seniority in the form of retirement age. Consider the ideal scenario: a woman who started her career at 18 and worked continuously until she was 60. Considering that men have a later retirement, and the fact that very few people work continuously and do not start at the age of 18, it is possible to obtain an adequate boundary for the maximum possible length of service.

In [51]:
print('The maximum possible length of seniority is:', (60 - 18) * 220, 'дней')

The maximum possible length of seniority is: 9240 дней


Let's replace the abnormally large values of seniority according to the following logic: calculate the median value of seniority for each value of age in the sample of clients whose seniority value is less than or equal to 9240 days and replace them with abnormally long seniority.

In [52]:
zf = df[df['days_employed'] <= 9240] # formation of a new dataframe with a selection of clients,
# who have experience less than or equal to 9240 days
for days in df['dob_years'].unique(): # cycle formation by unique age values
    median = zf.loc[df['dob_years'] == days,
                    'days_employed'].median() # calculating the median for the 'days_employed' column in the sample
    df.loc[(df['days_employed'] > 9240) &
           (df['dob_years'] == days), 'days_employed'] = median # replacement of abnormally long experience

Let's check by displaying the first ten rows of the table:

In [53]:
df.head(10) # check

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0    1816.843492         53   среднее             1   
5         0     926.185831         27    высшее             0   
6         0    2879.202052         43    высшее             0   
7         0     152.779569         50   СРЕДНЕЕ             1   
8         2    6929.865299         35    ВЫСШЕЕ             0   
9         0    2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0     253875.64   
1   женат / замужем                 0      F   сотрудник     0     112080.01   
2   женат / замужем                 0      M   сотрудник     0     145885.95   
3   женат / замужем                 0      M   сотрудник     0     267628.55   
4  гражданский брак                 1      F   пенсионер     0     158616.08   
5  гражданский брак                 1      M   компаньон     0     255763.57   
6   женат / замужем                 0      F   компаньон     0     240525.97   
7   женат / замужем                 0      M   сотрудник     0     135823.93   
8  гражданский брак                 1      F   сотрудник     0      95856.83   
9   женат / замужем                 0      M   сотрудник     0     144425.94   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

### Changing data types

Let's change the data type in the `total_income` column from real to integer, this will save the table from visual congestion, and the accuracy of the calculations will not suffer.

In [54]:
df['days_employed'] = df['days_employed'].astype('int') # replacing the float data type in the 'days_employed' column 
# with an integer

Let's check by displaying the first five rows of the table:

In [55]:
df.head() # getting first 5 rows of df table

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   Среднее             1   
3         3           4124         32   среднее             1   
4         0           1816         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0     253875.64   
1   женат / замужем                 0      F   сотрудник     0     112080.01   
2   женат / замужем                 0      M   сотрудник     0     145885.95   
3   женат / замужем                 0      M   сотрудник     0     267628.55   
4  гражданский брак                 1      F   пенсионер     0     158616.08   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

### Removing duplicates

First, let's determine if there are explicit duplicates in the table. Use the `duplicated()` method in conjunction with `sum()` as we need to see which columns have duplicates in the data.

In [56]:
df.duplicated().sum() # counting explicit duplicates

55

Let's display the first few rows of the table with repeated rows for review.

In [57]:
df[df.duplicated()].head() # getting the first 5 rows of a table with duplicate rows

children  days_employed  dob_years education  education_id  \
2849         0           1574         41   среднее             1   
4182         1           1574         34    ВЫСШЕЕ             0   
4851         0           2090         60   среднее             1   
5557         0           1860         58   среднее             1   
7808         0           1574         57   среднее             1   

         family_status  family_status_id gender income_type  debt  \
2849   женат / замужем                 0      F   сотрудник     0   
4182  гражданский брак                 1      F   сотрудник     0   
4851  гражданский брак                 1      F   пенсионер     0   
5557  гражданский брак                 1      F   пенсионер     0   
7808  гражданский брак                 1      F   пенсионер     0   

      total_income                  purpose  
2849     142594.40  покупка жилья для семьи  
4182     142594.40                  свадьба  
4851     118514.49                  свадьба  
5557     118514.49          сыграть свадьбу  
7808     118514.49    на проведение свадьбы

An analysis of a table with repeating rows allows us to conclude that there are matches mainly in the number of days of work experience and the amount of monthly income. Such duplication of data is possible both due to manual filling and problems when importing data from third-party resources. Let's remove duplicate rows and change the indexing of rows.

In [58]:
df = df.drop_duplicates().reset_index(drop=True) # removal of duplicate rows 
# (with removal of old indexes and formation of new ones)

Make sure to get rid of duplicate rows in the table.

In [59]:
df.duplicated().sum() # checking for duplicate lines

0

The `education` column contains data written using letters of different case. First, let's parse this column using the `value_counts()` method.

In [60]:
df['education'].value_counts() # counting implicit duplicates

среднее                13704
высшее                  4710
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   273
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64

<div class="alert alert-block alert-info">
<b>Note:</b>
    
    The 'education' column shows the levels of education in Russian:
    
    * начальное - Elementary School
    * среднее - Middle School
    * неоконченное высшее - Some University Courses
    * высшее - Master’s Degree
    * ученая степень - Doctorate Degree
</div>

Let's fix the different spelling of education data, bringing everything to the same register. Such spelling errors are common when entering data manually by different people.

In [61]:
df['education'] = df['education'].str.lower() # converting all characters in the 'education' column to lowercase
df['education'].value_counts() # check

среднее                15187
высшее                  5251
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

Additionally, we check the dataset by displaying the first ten lines on the screen.

In [62]:
df.head(10) # check

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0           1816         53   среднее             1   
5         0            926         27    высшее             0   
6         0           2879         43    высшее             0   
7         0            152         50   среднее             1   
8         2           6929         35    высшее             0   
9         0           2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0     253875.64   
1   женат / замужем                 0      F   сотрудник     0     112080.01   
2   женат / замужем                 0      M   сотрудник     0     145885.95   
3   женат / замужем                 0      M   сотрудник     0     267628.55   
4  гражданский брак                 1      F   пенсионер     0     158616.08   
5  гражданский брак                 1      M   компаньон     0     255763.57   
6   женат / замужем                 0      F   компаньон     0     240525.97   
7   женат / замужем                 0      M   сотрудник     0     135823.93   
8  гражданский брак                 1      F   сотрудник     0      95856.83   
9   женат / замужем                 0      M   сотрудник     0     144425.94   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Let's check the other columns for implicit duplicates.

In [63]:
df['family_status'].value_counts() # counting implicit duplicates in the 'family_status' column

женат / замужем          12344
гражданский брак          4162
Не женат / не замужем     2810
в разводе                 1195
вдовец / вдова             959
Name: family_status, dtype: int64

<div class="alert alert-block alert-info">
<b>Note:</b>
    
    The 'family_status' column shows the marital status in Russian:
    
    * женат / замужем - married
    * гражданский брак - civil marriage
    * не женат / не замужем - single
    * в разводе - divorced
    * вдовец / вдова - widower / widow
</div>

Here you can correct the spelling of the `Не женат / не замужем` (`Single`) category in lowercase letters, in a common style with the rest of the data in this column.

In [64]:
df['family_status'] = df['family_status'].replace('Не женат / не замужем','не женат / не замужем') # fix alternative category
# value in 'family_status' column
df['family_status'].value_counts() # check

женат / замужем          12344
гражданский брак          4162
не женат / не замужем     2810
в разводе                 1195
вдовец / вдова             959
Name: family_status, dtype: int64

In [65]:
df['gender'].value_counts() # count implicit duplicates in 'gender' column

F      14188
M       7281
XNA        1
Name: gender, dtype: int64

Output the entire string with the value `XNA`.

In [66]:
df[df['gender'] == 'XNA']

children  days_employed  dob_years            education  education_id  \
10690         0           2358         24  неоконченное высшее             2   

          family_status  family_status_id gender income_type  debt  \
10690  гражданский брак                 1    XNA   компаньон     0   

       total_income               purpose  
10690     203905.16  покупка недвижимости

This did not clarify the situation, let's make a decision to replace `'XNA'` with `'M'`, since this value occurs only once in 21525 values and this will not affect the accuracy of calculations in the future.

In [67]:
df['gender'] = df['gender'].replace('XNA', 'M') # replacing 'XNA' with 'M'
df.loc[10690, 'gender'] # check

'M'

In [68]:
df['income_type'].value_counts() # count implicit duplicates in 'income_type' column

сотрудник          11090
компаньон           5080
пенсионер           3837
госслужащий         1457
безработный            2
предприниматель        2
студент                1
в декрете              1
Name: income_type, dtype: int64

<div class="alert alert-block alert-info">
<b>Note:</b>
    
    The 'income_type' column shows the type of employment in Russian:
    
    * сотрудник - employee
    * компаньон - companion
    * пенсионер - pensioner
    * госслужащий - civil servant
    * безработный - unemployed
    * предприниматель - entrepreneur
    * студент - student
    * в декрете - on decree
</div>

Here the data is in order and does not require any adjustments.

In [69]:
df['purpose'].value_counts() # count implicit duplicates in 'purpose' column

свадьба                                   793
на проведение свадьбы                     772
сыграть свадьбу                           769
операции с недвижимостью                  675
покупка коммерческой недвижимости         662
покупка жилья для сдачи                   652
операции с жильем                         652
операции с коммерческой недвижимостью     650
покупка жилья                             646
жилье                                     646
покупка жилья для семьи                   638
строительство собственной недвижимости    635
недвижимость                              633
операции со своей недвижимостью           627
строительство жилой недвижимости          625
покупка недвижимости                      621
покупка своего жилья                      620
строительство недвижимости                619
ремонт жилью                              607
покупка жилой недвижимости                606
на покупку своего автомобиля              505
заняться высшим образованием      

<div class="alert alert-block alert-info">
<b>Note:</b>    
    The 'purpose' column shows the loan purposes in Russian, a note with the translation will be further in the section "Categorization of the purposes of the loan."
</div>

In my opinion, there are also implicit duplicates here. Similar loan purposes are written differently. You can reduce the number of loan purposes by combining similar ones, which I originally did. But the categorization of this type of data is present further in the project, so at this stage nothing needs to be done with this data.

### Creation of additional dataframes of dictionaries, decomposition of the original dataframe

Let's create two new dataframes in which:
* each unique value from `education` corresponds to a unique value `education_id` - in the first one;
* each unique value from `family_status` corresponds to a unique value `family_status_id` - in the second.

In [70]:
education_log = df[['education','education_id']] # creating a dataframe in which each unique value from 'education'
# matches the unique value of 'education_id'
family_status_log = df[['family_status','family_status_id']] # creation of a dataframe in which each unique value
# from 'family_status' matches the unique value of 'family_status_id'

Let's see what happens by displaying the first five lines of each dataframe.

In [71]:
education_log.head() # getting the first 5 rows of the 'education_log' dataframe

education  education_id
0    высшее             0
1   среднее             1
2   среднее             1
3   среднее             1
4   среднее             1

<div class="alert alert-block alert-info">
<b>Note:</b>
    
    The 'education' column shows the levels of education in Russian:
    
    * начальное - Elementary School
    * среднее - Middle School
    * неоконченное высшее - Some University Courses
    * высшее - Master’s Degree
    * ученая степень - Doctorate Degree
</div>

In [72]:
family_status_log.head() # getting the first 5 rows of the 'family_status_log' dataframe

family_status  family_status_id
0   женат / замужем                 0
1   женат / замужем                 0
2   женат / замужем                 0
3   женат / замужем                 0
4  гражданский брак                 1

<div class="alert alert-block alert-info">
<b>Note:</b>
    
    The 'family_status' column shows the marital status in Russian:
    
    * женат / замужем - married
    * гражданский брак - civil marriage
    * не женат / не замужем - single
    * в разводе - divorced
    * вдовец / вдова - widower / widow
</div>

Let's remove duplicates from the received dataframes.

In [73]:
education_log = education_log.drop_duplicates().reset_index(drop=True) # removing duplicates from 'education log' dataframe
family_status_log = family_status_log.drop_duplicates().reset_index(drop=True) # removing duplicates from dataframe 'family_status_log'
display(education_log.head()) # check
family_status_log.head() # check

education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  не женат / не замужем                 4

Let's remove the `education` and `family_status` columns from the original dataframe, leaving only their identifiers: `education_id` and `family_status_id`.

In [74]:
df = df.drop(columns=['education', 'family_status']) # removing columns 'education' and 'family_status'
df.head() # check

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0           1816         53             1                 1      F   

  income_type  debt  total_income                     purpose  
0   сотрудник     0     253875.64               покупка жилья  
1   сотрудник     0     112080.01     приобретение автомобиля  
2   сотрудник     0     145885.95               покупка жилья  
3   сотрудник     0     267628.55  дополнительное образование  
4   пенсионер     0     158616.08             сыграть свадьбу

### Income categorization

Let's create a `total_income_category` column, in which we will assign categories depending on the size of the client’s income:
* 0–30000 — 'E';
* 30001–50000 — 'D';
* 50001–200000 — 'C';
* 200001–1000000 — 'B';
* 1000001 и выше — 'A'.

First, let's create a function.

In [75]:
def income_group(income): # returns the corresponding income group
    if income <= 30000:
        return 'E'
    if income <= 50000:
        return 'D'
    if income <= 200000:
        return 'C'
    if income <= 1000000:
        return 'B'
    return 'A'

Let's create a new column `total_income_category`.

In [76]:
df['total_income_category'] = df['total_income'].apply(income_group) # creating a new column
df.head() # check

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0           1816         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0     253875.64               покупка жилья   
1   сотрудник     0     112080.01     приобретение автомобиля   
2   сотрудник     0     145885.95               покупка жилья   
3   сотрудник     0     267628.55  дополнительное образование   
4   пенсионер     0     158616.08             сыграть свадьбу   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C

### Categorization of loan purposes

Let's create a `purpose_category` column, in which we will assign categories depending on the purpose of obtaining a loan **(Russian (English)**:
* 'операции с автомобилем' (*'car operations'*);
* 'операции с недвижимостью' (*'real estate transactions'*);
* 'проведение свадьбы' (*'holding a wedding'*);
* 'получение образования' (*'getting an education'*).

First, let's create a function.

In [77]:
def purpose_group(purpose): # returns the corresponding income group
    if 'свадьб' in purpose:
        return 'проведение свадьбы'
    if 'образовани' in purpose:
        return 'получение образования'
    if 'автомобил' in purpose: 
        return 'операции с автомобилем'
    return 'операции с недвижимостью'

Let's create a new column `purpose_category`.

In [78]:
df['purpose_category'] = df['purpose'].apply(purpose_group) # creating a new column
df.head() # check

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0           1816         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0     253875.64               покупка жилья   
1   сотрудник     0     112080.01     приобретение автомобиля   
2   сотрудник     0     145885.95               покупка жилья   
3   сотрудник     0     267628.55  дополнительное образование   
4   пенсионер     0     158616.08             сыграть свадьбу   

  total_income_category          purpose_category  
0                     B  операции с недвижимостью  
1                     C    операции с автомобилем  
2                     C  операции с недвижимостью  
3                     B     получение образования  
4                     C        проведение свадьбы

Additionally, we will check the result of data categorization by loan purposes.

In [79]:
df['purpose_category'].value_counts() # check

операции с недвижимостью    10814
операции с автомобилем       4308
получение образования        4014
проведение свадьбы           2334
Name: purpose_category, dtype: int64

<div class="alert alert-block alert-info">
<b>Note:</b>
    
    The 'purpose_category' column shows the loan purpose categories in Russian:
    
    * операции с автомобилем - car operations
    * операции с недвижимостью - real estate transactions
    * проведение свадьбы - holding a wedding
    * получение образования - getting an education
</div>

**Conclusions**

Preprocessing found several problems in the data:

- violations in the display of data (use of letters of different case, the number of decimal places, negative values, abnormally large values),
- missing values,
- duplicates - explicit and implicit.

Having corrected the problems, we simplified the work with the table, which in the future will help to make the study more accurate.

Work has also been done on data categorization and dataframe decomposition, which will help simplify hypothesis testing.

## Hypothesis testing

### Dependence of loan repayment on time on the number of children

It is necessary to compare information on the number of children in the family and the presence of debt on loans. To do this, we will form a pivot table according to the following principle - the first column `children` will display the number of children in the family, the second column `children len` will display the number of families with the current number of children, the third column `debt mean` will display the proportion families that are in debt with the current number of children (arithmetic average of the data from the 'debt' column), the fourth column `debt sum` will display the number of families indebted with the current number of children.

In [80]:
import numpy as np # importing the numpy library

In [81]:
df.pivot_table(index='children', aggfunc={'children': len, 'debt': [np.mean, np.sum]}) # compilation of a pivot table

children      debt        
              len      mean     sum
children                           
0           14106  0.075358  1063.0
1            4856  0.091639   445.0
2            2128  0.094925   202.0
3             330  0.081818    27.0
4              41  0.097561     4.0
5               9  0.000000     0.0

### Conclusion on the first hypothesis

Based on the results of testing the first hypothesis, we see that with the maximum number of children in the family (5), there are no loan debts. In groups with 1 to 4 children, the percentage of families with loan debt varies from 8.18% to 9.76%. In the absence of children, the percentage of families with loan debts is 7.54% of the group. This allows us to conclude that families with the number of children 0 and 5, as a rule, either do not have debts or try to repay them on time. Families with 1 to 4 children may be in default on loans.

### Dependence of loan repayment on time on marital status

It is necessary to compare information on marital status and the presence of debt on loans. To do this, we will form a pivot table according to the following principle - in the first column `family_status_id` we will display marital status, in the second column `debt mean` we will display the proportion of customers who have debts with the current family status (arithmetic average of the data from the 'debt '), the third column `debt sum` will display the number of clients who have debts with the current family status, the fourth column `family_status_id len` will display the number of clients with the current family status. Keep in mind that in previous iterations, the `family_status` column was removed from the original dataframe.

In [82]:
df.pivot_table(index='family_status_id', aggfunc={'family_status_id': len,
                                                  'debt': [np.mean, np.sum]}) # compilation of a pivot table

debt        family_status_id
                      mean    sum              len
family_status_id                                  
0                 0.075421  931.0            12344
1                 0.093224  388.0             4162
2                 0.065693   63.0              959
3                 0.071130   85.0             1195
4                 0.097509  274.0             2810

In [85]:
family_status_log # cheat sheet

family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  не женат / не замужем                 4

<div class="alert alert-block alert-info">
<b>Note:</b>
    
    The 'family_status' column shows the marital status in Russian:
    
    * женат / замужем - married
    * гражданский брак - civil marriage
    * не женат / не замужем - single
    * в разводе - divorced
    * вдовец / вдова - widower / widow
</div>

### Conclusion on the second hypothesis

Based on the results of testing the second hypothesis, we see that the largest number of debts is owned by unmarried and civil-married clients, 9.75% and 9.32%, respectively. Married and divorced clients average 7.54% and 7.11%. Widowers/widows are the most careful about their debts, they are only 6.57% of the group.

### The dependence of the loan repayment on time on the level of income

It is necessary to compare information on the level of income and the presence of debt on loans. To do this, we will form a pivot table according to the following principle - in the first column `total_income_category` we will display the client's income category, the second column `debt mean` will display the proportion of clients who have debts at the current income category (arithmetic average of the data from the column ' debt'), the third column `debt sum` will display the number of clients who have debts for the current income category, the fourth column `total_income_category len` will display the number of clients from the current income category. At the same time, it should be remembered that in previous iterations, customers were categorized by income level.

In [84]:
df.pivot_table(index='total_income_category', aggfunc={'total_income_category': len,
                                                       'debt': [np.mean, np.sum]}) # compilation of a pivot table

debt         total_income_category
                           mean     sum                   len
total_income_category                                        
A                      0.080000     2.0                    25
B                      0.070607   356.0                  5042
C                      0.084836  1360.0                 16031
D                      0.060000    21.0                   350
E                      0.090909     2.0                    22

Cheat sheet by income category:
* 0–30000 — 'E';
* 30001–50000 — 'D';
* 50001–200000 — 'C';
* 200001–1000000 — 'B';
* 1000001 и выше — 'A'.

### Conclusion on the third hypothesis

According to the results of testing the third hypothesis, we see that the largest number of delays is allowed by customers with a minimum income level (0-30000), they are 9.09% of the group. Clients with maximum (1000001 and above) and average income (50001-200000) have an average of 8.48% and 8.00%. Clients with an income of "above average" (200001-1000000) and "below average" (30001-50000) are the most accurate in terms of loan repayments on time, their rates are 7.06% and 6.00% respectively.

### Influence of different goals on loan repayment on time

It is necessary to compare information on the purpose of the loan and the presence of debt on loans. To do this, a pivot table according to the following principle - in the first column `purpose_category` we will display the purpose of the client's loan, in the second column `debt mean` the proportion of customers who have debts with the current purpose of the loan will be displayed (arithmetic average of the data from the column 'debt '), the third column `debt sum` will display the number of customers who have debts for the current purpose of the loan, the fourth column `purpose_category len` will show the number of customers with the current purpose of the loan. At the same time, it should be remembered that in previous iterations, customers were categorized according to the purpose of the loan.

In [86]:
df.pivot_table(index='purpose_category', aggfunc={'purpose_category': len,
                                                  'debt': [np.mean, np.sum]}) # compilation of a pivot table

debt        purpose_category
                              mean    sum              len
purpose_category                                          
операции с автомобилем    0.093547  403.0             4308
операции с недвижимостью  0.072314  782.0            10814
получение образования     0.092177  370.0             4014
проведение свадьбы        0.079692  186.0             2334

<div class="alert alert-block alert-info">
<b>Note:</b>
    
    The 'purpose_category' column shows the loan purpose categories in Russian:
    
    * операции с автомобилем - car operations
    * операции с недвижимостью - real estate transactions
    * проведение свадьбы - holding a wedding
    * получение образования - getting an education
</div>

### Conclusion on the fourth hypothesis

According to the results of testing the fourth hypothesis, we see that the largest number of delinquencies are allowed by customers who plan to take loans for car operations and education, they are 9.35% and 9.22% of the groups. Clients planning to take out a loan for real estate transactions and weddings try to avoid delays, their rates are 7.23% and 7.97%, respectively.

## Research results

I tested four hypotheses and found:

1. The number of children affects the loan repayment term.
2. Marital status affects the loan repayment period.
3. The level of income affects the loan repayment period.
4. The purpose of the loan affects the repayment term of the loan.

All hypotheses were confirmed.
However, during the audit it turned out that it is not worth making unambiguous conclusions on the degree of solvency of the client based on the data provided. For example, clients with a large number of children are more responsible for paying off debts, but clients with a high level of income, on the contrary. The study also led to the conclusion that the stereotype regarding the low solvency of clients who take out a loan for the purpose of holding a wedding is untenable.

Various customer data allow you to make reasonable forecasts of solvency. In the development of this topic, in the future, you can assign subcategories of clients that take into account the totality of data, on which, in fact, the scoring calculation methods are based.